File: leetcode/C++/triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int cur[200]={0},next[200], sol=10000;
        for(int i=0; i<triangle.size(); ++i){
            for(int j=0; j<triangle[i].size(); ++j){
                if(j==0) next[j]=cur[j]+triangle[i][j];
                else if(j+1==triangle[i].size()) next[j]=cur[j-1]+triangle[i][j];
                else next[j]=triangle[i][j]+min(cur[j-1],cur[j]);
            }
            for(int j=0; j<triangle[i].size(); ++j) cur[j]=next[j];
        }
        for(int i=0; i<triangle.size(); ++i)
            if(sol>cur[i]) sol=cur[i];
        return sol;
    }
};

File: leetcode/Python/triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]):
        cur = [0 for i in range(200)]
        next = [0 for i in range(200)]
        sol = 10000
        for i, row in enumerate(triangle):
            for j, v in enumerate(row):
                if j == 0:
                    next[j]=cur[j]+v
                elif j + 1 == len(row):
                    next[j]=cur[j-1]+v
                else:
                    next[j]=min(cur[j-1],cur[j])+v
            cur=next.copy()
        for i in range(len(triangle)):
            if sol > cur[i]:
                sol = cur[i]
        return sol